### ISSUE TITLE
eBPF 1% of inbound connections stall on create

### ISSUE THREAD

**Issue by salanki** [projectcalico/calico#4431]
When running Calico in eBPF mode, 0.1% to 1% of connections stall when created towards a ServiceIP / LoadBalancer IP of certain applications. Accessing the Pod IPs directly leads to no stalls, neither does switching back service proxy to `kube-proxy`. The connection is successfully opened (SYN-ACK) but no further data is received.

This behavior has been proven with the following applications:
- traefik
- istio

## Expected Behavior
All connections should succeed.

## Current Behavior
```
hey -z 48h -t 20 -c 1 http://my-traefik-service

Summary:
  Total:        44.7973 secs
  Slowest:      0.0555 secs
  Fastest:      0.0275 secs
  Average:      0.0298 secs
  Requests/sec: 3.6386

  Total data:   2093 bytes
  Size/request: 13 bytes

Response time histogram:
  0.028 [1]     |
  0.030 [124]   |■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.033 [32]    |■■■■■■■■■■
  0.036 [3]     |■
  0.039 [0]     |
  0.042 [0]     |
  0.044 [0]     |
  0.047 [0]     |
  0.050 [0]     |
  0.053 [0]     |
  0.055 [1]     |


Latency distribution:
  10% in 0.0285 secs
  25% in 0.0286 secs
  50% in 0.0290 secs
  75% in 0.0300 secs
  90% in 0.0325 secs
  95% in 0.0330 secs
  99% in 0.0555 secs

Details (average, fastest, slowest):
  DNS+dialup:   0.0000 secs, 0.0275 secs, 0.0555 secs
  DNS-lookup:   0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:    0.0001 secs, 0.0000 secs, 0.0015 secs
  resp wait:    0.0291 secs, 0.0273 secs, 0.0544 secs
  resp read:    0.0003 secs, 0.0000 secs, 0.0025 secs

Status code distribution:
  [200] 161 responses

Error distribution:
  [2]   Get http://64.79.105.221: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

## Steps to Reproduce (for bugs)
1. Run traefik in a Pod on a node with calico eBPF service proxy
2. Use hey or similar tool to launch a bunch of sequential requests to a Service IP of that pod
3. See how a few connections stall and time out

## Your Environment

* Calico version: 3.16.0 and 3.18.0
* Orchestrator version (e.g. kubernetes, mesos, rkt): Kubernetes 1.18.0
* Operating System and version: Linux g4efbf7 5.4.0-54-generic #60~18.04.1-Ubuntu SMP Fri Nov 6 17:25:16 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
* Link to your project (optional): http://www.coreweave.com

**Comment by lmm** [comment]
FYI @tomastigera

**Comment by tomastigera** [comment]
@salanki  do you see any unexpected RSTs? Possibly related to https://github.com/projectcalico/calico/issues/4290

**Comment by salanki** [comment]
Do not see any RSTs. The connection just stalls after the SYN-ACK.

**Comment by bcreane** [comment]
cc @fasaxc , @tomastigera - sounds like the initial question of RSTs has been answered. Any other hypotheses that we can ask @salanki to take a look for?

**Comment by fasaxc** [comment]
@salanki Where are you running `hey`?  In a pod, on the host with Traefik?  On an external host?

**Comment by salanki** [comment]
External host, outside of the cluster.

**Comment by fasaxc** [comment]
Do you see the same with a super simple service, such as an `nginx` pod?

**Comment by salanki** [comment]
Let me run another test with Nginx. I have tested many different applications (Istio, Traefik) but all of them are written in Go. Would be interesting to try something written in C.

**Comment by fasaxc** [comment]
@salanki Can you give more details of how to set up Traefik, for example, in order to repro this.  I thought Traefik was an ingress controller so what exactly is the set up that you're using and what is Traefik providing ingress for?

**Comment by alloran** [comment]
Got same issue with nodePorts/externalIPs of services having more than one endpoint.


Service:
```
NAME      TYPE       CLUSTER-IP       EXTERNAL-IP     PORT(S)        AGE
nginx     NodePort   10.235.127.158   10.238.164.34   80:31471/TCP   144m
```

hey running on the external vm in the same subnet as all k8s nodes.
```
# hey -disable-keepalive  -z 60s  http://10.238.164.34/

Summary:
  Total:	73.9967 secs
  Slowest:	20.0002 secs
  Fastest:	0.0003 secs
  Average:	0.0017 secs
  Requests/sec:	6921.9191
  
  Total data:	313391736 bytes
  Size/request:	612 bytes

Response time histogram:
  0.000 [1]	|
  2.000 [512075]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  4.000 [0]	|
  6.000 [0]	|
  8.000 [0]	|
  10.000 [0]	|
  12.000 [0]	|
  14.000 [0]	|
  16.000 [0]	|
  18.000 [0]	|
  20.000 [2]	|


Latency distribution:
  10% in 0.0007 secs
  25% in 0.0010 secs
  50% in 0.0012 secs
  75% in 0.0017 secs
  90% in 0.0027 secs
  95% in 0.0037 secs
  99% in 0.0066 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0005 secs, 0.0003 secs, 20.0002 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0001 secs, 0.0000 secs, 0.0088 secs
  resp wait:	0.0009 secs, 0.0001 secs, 0.2094 secs
  resp read:	0.0002 secs, 0.0000 secs, 19.9970 secs

Status code distribution:
  [200]	512078 responses

Error distribution:
  [121]	Get "http://10.238.164.34/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

```

Calico 3.18.1, ebpf dataplane, vxlan overlay, k8s 1.20.4, ubuntu 18.04 with 5.4 kernel

No errors in ipvs mode.

Also, ebpf dataplane does not honor `externalTrafficPolicy: Local` for externalIPs, or maybe I misunderstood  something?

```
#calico-node -bpf nat dump
...
10.238.164.34 port 80 proto 6 id 12 count 2 local 1 #xeternalIP
	12:0	 10.235.136.5:80
	12:1	 10.235.211.4:80

10.238.164.3 port 31471 proto 6 id 14 count 1 local 0 #nodePort
	14:0	 10.235.211.4:80
10.238.164.4 port 31471 proto 6 id 12 count 1 local 1 #nodePort
	12:0	 10.235.136.5:80

```
I've also tried `fasaxc/calico-node:ct-syns-2` image with same result, seems like this issue is not related to #4290

**Comment by fasaxc** [comment]
@alloran can you give this image a try: `calico/node:v3.19.0-0.dev-121-gf2ea8a855295`

That's the latest cut of v3.19.  It has some more fixes for external IPs.

**Comment by alloran** [comment]
@fasaxc 
Tested with `calico/node:v3.19.0-0.dev-121-gf2ea8a855295`, same result

3.19, single nginx pod
```
Summary:
  Total:	60.0054 secs
  Slowest:	0.0218 secs
  Fastest:	0.0012 secs
  Average:	0.0062 secs
  Requests/sec:	8079.4499
  
  Total data:	296704332 bytes
  Size/request:	612 bytes

Response time histogram:
  0.001 [1]	|
  0.003 [866]	|
  0.005 [198000]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.007 [218379]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.009 [17051]	|■■■
  0.012 [41242]	|■■■■■■■■
  0.014 [6301]	|■
  0.016 [2535]	|
  0.018 [311]	|
  0.020 [96]	|
  0.022 [29]	|


Latency distribution:
  10% in 0.0048 secs
  25% in 0.0050 secs
  50% in 0.0056 secs
  75% in 0.0064 secs
  90% in 0.0098 secs
  95% in 0.0108 secs
  99% in 0.0126 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0007 secs, 0.0012 secs, 0.0218 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0109 secs
  resp wait:	0.0053 secs, 0.0003 secs, 0.0203 secs
  resp read:	0.0001 secs, 0.0000 secs, 0.0138 secs

Status code distribution:
  [200]	484811 responses


```
3.19, two nginx pods
```
Summary:
  Total:	79.7519 secs
  Slowest:	20.0002 secs
  Fastest:	0.0009 secs
  Average:	0.0029 secs
  Requests/sec:	4629.4847
  
  Total data:	225881856 bytes
  Size/request:	612 bytes

Response time histogram:
  0.001 [1]	|
  2.001 [369085]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  4.001 [0]	|
  6.001 [0]	|
  8.001 [0]	|
  10.001 [0]	|
  12.000 [0]	|
  14.000 [0]	|
  16.000 [0]	|
  18.000 [0]	|
  20.000 [2]	|


Latency distribution:
  10% in 0.0016 secs
  25% in 0.0018 secs
  50% in 0.0020 secs
  75% in 0.0027 secs
  90% in 0.0045 secs
  95% in 0.0068 secs
  99% in 0.0125 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0010 secs, 0.0009 secs, 20.0002 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0000 secs, 0.0000 secs, 0.0082 secs
  resp wait:	0.0016 secs, 0.0002 secs, 0.2068 secs
  resp read:	0.0002 secs, 0.0000 secs, 19.9977 secs

Status code distribution:
  [200]	369088 responses

Error distribution:
  [122]	Get "http://10.238.164.34/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)

```

**Comment by salanki** [comment]
Now that this is reproduced, any ideas on how to mitigate it?

**Comment by salanki** [comment]
We still have this issue running latest version of Calico and latest 5.13 mainline kernel as well as Ubuntu kernels. This is reproducible in VMs as well as on bare metal.

**Comment by salanki** [comment]
This issue regretfully still exists as of today (Felix HEAD https://github.com/projectcalico/felix/commit/012f8389275c2c55257de87c8306f020a53ed56e).

This is the manifest we are using to reproduce:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx-deployment
  labels:
    app: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.14.2
        ports:
        - containerPort: 80
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
    - port: 80
      name: http
  selector:
    app: nginx
  type: LoadBalancer
  externalTrafficPolicy: Local
```

```
$ hey -disable-keepalive  -z 60s  http://10.135.208.6/

Summary:
  Total:	75.6985 secs
  Slowest:	1.0014 secs
  Fastest:	0.0005 secs
  Average:	0.0030 secs
  Requests/sec:	17487.5167

  Total data:	810151524 bytes
  Size/request:	810 bytes

Response time histogram:
  0.000 [1]	|
  0.101 [999995]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.201 [2]	|
  0.301 [0]	|
  0.401 [0]	|
  0.501 [0]	|
  0.601 [0]	|
  0.701 [0]	|
  0.801 [0]	|
  0.901 [0]	|
  1.001 [2]	|


Latency distribution:
  10% in 0.0017 secs
  25% in 0.0018 secs
  50% in 0.0021 secs
  75% in 0.0025 secs
  90% in 0.0029 secs
  95% in 0.0032 secs
  99% in 0.0061 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0006 secs, 0.0005 secs, 1.0014 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0001 secs, 0.0000 secs, 0.0481 secs
  resp wait:	0.0020 secs, 0.0001 secs, 0.0927 secs
  resp read:	0.0002 secs, 0.0000 secs, 0.0528 secs

Status code distribution:
  [200]	1000000 responses

Error distribution:
  [2]	Get http://10.135.208.6/: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
```

The `LoadBalancer` IP is allocated via `MetalLB` and announced by Calico BGP.

**Comment by salanki** [comment]
@tomastigera: I can regretfully confirm that this issue still exists with https://github.com/projectcalico/felix/pull/2983. Tested via: calico/node:v3.21.0-0.dev-121-g2004f2a17a2.

```
$ hey -disable-keepalive  -z 60s -t 1  http://10.135.208.6/

Summary:
  Total:	60.0027 secs
  Slowest:	0.0241 secs
  Fastest:	0.0003 secs
  Average:	0.0029 secs
  Requests/sec:	28774.3879

  Total data:	1056600252 bytes
  Size/request:	1056 bytes

Response time histogram:
  0.000 [1]	|
  0.003 [780832]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  0.005 [205750]	|■■■■■■■■■■■
  0.007 [10612]	|■
  0.010 [2051]	|
  0.012 [481]	|
  0.015 [190]	|
  0.017 [45]	|
  0.019 [27]	|
  0.022 [9]	|
  0.024 [2]	|


Latency distribution:
  10% in 0.0005 secs
  25% in 0.0006 secs
  50% in 0.0014 secs
  75% in 0.0026 secs
  90% in 0.0031 secs
  95% in 0.0036 secs
  99% in 0.0054 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0006 secs, 0.0003 secs, 0.0241 secs
  DNS-lookup:	0.0000 secs, 0.0000 secs, 0.0000 secs
  req write:	0.0002 secs, 0.0000 secs, 0.0159 secs
  resp wait:	0.0019 secs, 0.0001 secs, 0.0197 secs
  resp read:	0.0002 secs, 0.0000 secs, 0.0220 secs

Status code distribution:
  [200]	1000000 responses

Error distribution:
  [70]	Get "http://10.135.208.6/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

**Comment by salanki** [comment]
https://github.com/projectcalico/calico/issues/4712 seems related, however our issue does not occur with `externalTrafficPolicy: Cluster`.

**Comment by tomastigera** [comment]
So @salanki this commint causes the troubles for you, right? https://github.com/projectcalico/felix/pull/2686

**Comment by tomastigera** [comment]
#4310 may look similar, but uses 3.17 while the :arrow_up:  PR was merged in 3.19 :thinking:

**Comment by tomastigera** [comment]
The key problem in this is that an unrelated feature in our kube-proxy replacement overwrites data used by the cleaner that removes conntrack entries when a backend no longer exists. This clash only happens with the NodePort having ExternalTrafficPolicy set to Local and it shows the symptoms if there is more than a single backend. The result is that when we sweep the connections (every ~10s) we erroneously remove the conntrack entry of a live connection. Since there are 2 entries for NATed connections, the BPF logs show different symptoms depending on which of the 2 entries got removed first and when new packets on the connection pass through. Due to removing the conntrack entries, the connection stalls.

Note that LoadBalancers ultimately create NodePorts.

cc @seh
